File: hackerrank/problem solving/algorithms/2 - implementation/22 - angry professor/with morfo library/angry_professor.go
// https://www.hackerrank.com/challenges/angry-professor/problem?isFullScreen=true

package main

import (
	"fmt"
	"morfo/morfocollections"
	"morfo/morfoio"
)

func main() {
	n := morfoio.Read[int]()

	for range n {
		morfoio.SkipInput()
		threshold := morfoio.Read[int]()
		arrivalTimes := morfoio.ReadLn[int]()

		if angryProfessor(arrivalTimes, threshold) {
			fmt.Println("YES")
		} else {
			fmt.Println("NO")
		}
	}
}

// k: length of arrivalTimes
// T: O(k)
// S: O(1) extra space
func angryProfessor(arrivalTimes []int, threshold int) bool {
	onTime := morfocollections.CountFunc(arrivalTimes, func(t int) bool { return t <= 0 })
	return onTime < threshold
}

File: hackerrank/problem solving/algorithms/2 - implementation/22 - angry professor/with morfo library/angry_professor.py
# https://www.hackerrank.com/challenges/angry-professor/problem?isFullScreen=true

import morfo


def main():
    n = morfo.read(int)
    for _ in range(n):
        morfo.skip_input()
        threshold = morfo.read(int)
        arrival_times = morfo.readln(int, list)
        print("YES" if angry_professor(arrival_times, threshold) else "NO")


# k: length of arrival_times
# T: O(k)
# S: O(1) extra space
def angry_professor(arrival_times, threshold):
    on_time = sum(t <= 0 for t in arrival_times)
    return on_time < threshold


if __name__ == '__main__':
    main()
